## **Question 1**
- Install uv
- What's the version of uv you installed?
- Use --version to find out

In [5]:

!uv --version

uv 0.9.7 (0adb44480 2025-10-30)


## **Question 2**
- Use uv to install Scikit-Learn version 1.6.1
- What's the first hash for Scikit-Learn you get in the lock file?
- Include the entire string starting with sha256:, don't include quotes

In [7]:
!uv add scikit-learn==1.6.1

Using CPython 3.12.6 interpreter at: C:\Users\ELITEBOOK830\AppData\Local\Programs\Python\Python312\python.exe
Creating virtual environment at: .venv
Resolved 6 packages in 3.20s
Prepared 5 packages in 1m 32s
         If the cache and target directories are on different filesystems, hardlinking may not be supported.
         If this is intentional, set `export UV_LINK_MODE=copy` or use `--link-mode=copy` to suppress this warning.
Installed 5 packages in 3.21s
 + joblib==1.5.2
 + numpy==2.3.4
 + scikit-learn==1.6.1
 + scipy==1.16.3
 + threadpoolctl==3.6.0


sha256:b4fc2525eca2c69a59260f583c56a7557c6ccdf8deafdba6e060f94c1c59738e

## **Question 3**
Let's use the model!

- Write a script for loading the pipeline with pickle
- Score this record:

{

    "lead_source": "paid_ads",

    "number_of_courses_viewed": 2,

    "annual_income": 79276.0
    
}

What's the probability that this lead will convert?

- 0.333
- 0.533
- 0.733
- 0.933

In [8]:
import pickle

In [9]:
with open('pipeline_v1.bin', 'rb') as f:
    model = pickle.load(f)

record = {
    "lead_source": "paid_ads",
    "number_of_courses_viewed": 2,
    "annual_income": 79276.0
}

prob = model.predict_proba([record])[0, 1]

print(prob)


c:\Users\ELITEBOOK830\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\base.py:376: InconsistentVersionWarning: Trying to unpickle estimator DictVectorizer from version 1.6.1 when using version 1.5.1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations
  warnings.warn(


0.5336072702798061


c:\Users\ELITEBOOK830\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\base.py:376: InconsistentVersionWarning: Trying to unpickle estimator LogisticRegression from version 1.6.1 when using version 1.5.1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations
  warnings.warn(
c:\Users\ELITEBOOK830\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\base.py:376: InconsistentVersionWarning: Trying to unpickle estimator Pipeline from version 1.6.1 when using version 1.5.1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations
  warnings.warn(


## **Question 4**
Now let's serve this model as a web service

- Install FastAPI
- Write FastAPI code for serving the model
- Now score this client using requests:

url = "YOUR_URL"

client = {

    "lead_source": "organic_search",

    "number_of_courses_viewed": 4,

    "annual_income": 80304.0

}

requests.post(url, json=client).json()

What's the probability that this client will get a subscription?

- 0.334
- 0.534
- 0.734
- 0.934

In [ ]:
!pip install fastapi uvicorn

  Using cached fastapi-0.120.3-py3-none-any.whl.metadata (28 kB)
  Using cached uvicorn-0.38.0-py3-none-any.whl.metadata (6.8 kB)
  Using cached pydantic-2.12.3-py3-none-any.whl.metadata (87 kB)
   ---------------------------------------- 0.0/2.0 MB ? eta -:--:--
   ---------------------------------------- 0.0/2.0 MB ? eta -:--:--
   ---------------------------------------- 0.0/2.0 MB ? eta -:--:--
   ----- ---------------------------------- 0.3/2.0 MB ? eta -:--:--
   ----- ---------------------------------- 0.3/2.0 MB ? eta -:--:--
   ---------- ----------------------------- 0.5/2.0 MB 762.0 kB/s eta 0:00:02
   --------------- ------------------------ 0.8/2.0 MB 882.6 kB/s eta 0:00:02
   -------------------- ------------------- 1.0/2.0 MB 898.8 kB/s eta 0:00:02
   -------------------- ------------------- 1.0/2.0 MB 898.8 kB/s eta 0:00:02
   ------------------------- -------------- 1.3/2.0 MB 894.7 kB/s eta 0:00:01
   ------------------------- -------------- 1.3/2.0 MB 894.7 kB/s eta 

In [16]:
url = "http://127.0.0.1:8000/predict"
client = {
    "lead_source": "organic_search",
    "number_of_courses_viewed": 4,
    "annual_income": 80304.0
}
requests.post(url, json=client).json()

{'probability': 0.5340417283801275}

## **Question 5**
Download the base image agrigorev/zoomcamp-model:2025. You can easily make it by using docker pull command.

So what's the size of this base image?

- 45 MB
- 121 MB
- 245 MB
- 330 MB

You can get this information when running docker images - it'll be in the "SIZE" column.

In [3]:
!docker pull agrigorev/zoomcamp-model:2025

2025: Pulling from agrigorev/zoomcamp-model
816e9c0578b5: Pulling fs layer
799df5177ae2: Pulling fs layer
59e22667830b: Pulling fs layer
19f81082f6f0: Pulling fs layer
66cd357cf7a1: Pulling fs layer
760c2b9cc6ee: Pulling fs layer
816e9c0578b5: Download complete
66cd357cf7a1: Download complete
760c2b9cc6ee: Download complete
19f81082f6f0: Download complete
799df5177ae2: Download complete
59e22667830b: Download complete
Digest: sha256:14d79fde0bbf078eb18c99c2bd007205917b758ec11060b2994963a1e485c2ae
Status: Downloaded newer image for agrigorev/zoomcamp-model:2025
docker.io/agrigorev/zoomcamp-model:2025


In [5]:
!docker images

REPOSITORY                 TAG       IMAGE ID       CREATED       SIZE
agrigorev/zoomcamp-model   2025      14d79fde0bbf   10 days ago   181MB


## **Question 6**
Let's run your docker container!

After running it, score this client once again:

url = "YOUR_URL"

client = {

    "lead_source": "organic_search",

    "number_of_courses_viewed": 4,

    "annual_income": 80304.0

}

requests.post(url, json=client).json()
What's the probability that this lead will convert?

- 0.39
- 0.59
- 0.79
- 0.99

In [ ]:
url = "http://127.0.0.1:8000/predict"
client = {
    "lead_source": "organic_search",
    "number_of_courses_viewed": 4,
    "annual_income": 80304.0
}
requests.post(url, json=client).json()

{'probability': 0.5340417283801275}